File: sdk-objectivec/office365_odata_base/office365_odata_impl/Implementations/MSHttpConnection.m
/*******************************************************************************
 * Copyright (c) Microsoft Open Technologies, Inc.
 * All Rights Reserved
 * Licensed under the Apache License, Version 2.0.
 * See License.txt in the project root for license information.
 ******************************************************************************/

#import "MSHttpConnection.h"
#import "MSRequestImpl.h"
#import "MSResponseImpl.h"

@implementation MSHttpConnection

-(NSURLSessionDataTask *)execute:(id<MSRequest>)request :(void (^)(id<MSResponse> response, NSError *error))callback{
    
    NSMutableURLRequest* r = [(MSRequestImpl*)request getMutableRequest] ;
    
    NSURLSession *session = [NSURLSession sharedSession];
    NSURLSessionDataTask *task = [session dataTaskWithRequest:r completionHandler:^(NSData *data, NSURLResponse *response, NSError *error) {
        
        long statusCode = [(NSHTTPURLResponse*)response statusCode];
        
        if(statusCode < 200 || statusCode > 299){
            if(error == nil){
                NSArray * msj = [NSJSONSerialization JSONObjectWithData:data options: NSJSONReadingMutableContainers error:nil];
                error = [[NSError alloc] initWithDomain:@"Error in the Request" code:statusCode userInfo:(NSDictionary*)msj];
            }
            
            callback([[MSResponseImpl alloc] initWith:nil :(int)statusCode], error);
        }else{
            MSResponseImpl* responseImpl = [[MSResponseImpl alloc] initWith:data :(int)statusCode];
            callback(responseImpl, error);
        }
    }];
    
    return task;
}

-(id<MSRequest>)createRequest{
    return [[MSRequestImpl alloc] init];
}

@end

File: sdk-swift/office365_odata_impl/MSHttpConnection.swift
/*******************************************************************************
* Copyright (c) Microsoft Open Technologies, Inc.
* All Rights Reserved
* Licensed under the Apache License, Version 2.0.
* See License.txt in the project root for license information.
******************************************************************************/

import Foundation

public class MSHttpConnection : MSHttpTransport{

    required public init(){
    
    }
    
    public func execute(request: MSRequest, callback: ((response: MSResponse?, error: NSError?) -> Void?)) -> NSURLSessionDataTask {
        
        var r = (request as MSRequestImpl).getMutableRequest();
        
        var session = NSURLSession.sharedSession();
        var task = session.dataTaskWithRequest(r, completionHandler: { ( data : NSData!, response : NSURLResponse!, e : NSError! ) -> Void in
            
            var statusCode = (response as NSHTTPURLResponse).statusCode;
            
            if (statusCode < 200 || statusCode > 299){
            
                if(e == nil){
                    
                    var msj: AnyObject? = NSJSONSerialization.JSONObjectWithData(data,
                        options: NSJSONReadingOptions.MutableContainers, error: nil);
                    
                    var error = NSError(domain: "Error in the Request", code: statusCode, userInfo: msj as NSDictionary);
                    callback(response: MSResponseImpl(data: nil, status: statusCode), error: e);
                }
            }
            else{
            
                var responseImpl = MSResponseImpl(data: data,status: statusCode);
                callback(response: responseImpl, error: e);
            }
            
        });
        
        return task;
    }
    
    public func createRequest() -> MSRequest {
        return MSRequestImpl();
    }
}